# Data filtering

We will now see a couple of basic functions to filter and select a data object.

## filter()

**filter()** will filter **rows**.

If we take the **geneexp** objet as an example, we can filter only down-regulated genes using the **filter()** function of **{dplyr}**.

In case you need to import the data again:

```{r, eval=FALSE}
geneexp <- read_csv("DataViz_source_files-main/files/expression_20genes.csv")
```

```{r}
filter(geneexp, DE=="Down")
```

**==** is a **logical operator** that represens equality. It means that filter will return **rows in geneexp that are exactly equal to "Down"**.

Logical operators:

| Operator  | Description   |
|:--------: |:-------------:|
|     <     | less than    |
|     <=    | less than or equal to   |
|     >     | greater than  |
|     >=    | greater than or equal to|
|     ==    | exactly equal to   |
|     !=    | not equal to |
|    !x     | not x |
|    x | y  | x OR y |
|    x & y  | x AND y |

We can have several conditions.

For example, we may want to extract only **Up** or **Down** rows from geneexp column **DE**:

```{r}
filter(geneexp, DE=="Down" | DE=="Up")
```

Here, we introduce another operator, **|**, which mean **OR**, so rows will be kept if there is either **Down** or **Up** in **DE** column.

A good practice is to assign the filtered output to a new object, for example:

```{r}
geneexp_filt <- filter(geneexp, DE=="Down" | DE=="Up")
```

## select()

**select()** will select **columns**.

Select only columns **Gene** and **DE** from **geneexp**:

```{r}
select(geneexp, Gene, DE)
```

Again, a good practice is to save that subset into a new object:

```{r}
geneexp_selected <- select(geneexp, Gene, DE)
```

## rename()

You can easily rename a column using the **rename()** function.

For example, we can rename column **Gene** to **Symbol**, the following way:

```{r}
rename(geneexp, Symbol=Gene)
```

The first argument is the data, then comes the column(s) to rename in the format: *new_name=old_name*.


## Exercise 3

We will start from **gtf** object.

Read it as follows, if it is not in your environment:

```{r, eval=F}
gtf <- read_csv("DataViz_source_files-main/files/gencode.v44.annotation.csv")
```

1. Rename column **chr** to **Chromosome**. Assign to a new object.

<details>
<summary>
correction
</summary>

```{r}
gtf2 <- rename(gtf, chromosome=chr)
```

</details>

<br>
2. Keep only rows corresponding to the **+** strand and to chromosome **chr4**. assign to a new object.

<details>
<summary>
correction
</summary>

```{r}
gtf3 <- filter(gtf2, strand=="+" & chromosome=="chr4")
```

</details>

<br>

3. Select only columns **gene_symbol** and **gene_type**.

<details>
<summary>
correction
</summary>

```{r}
gtf4 <- select(gtf3, gene_symbol, gene_type)
```

</details>

<br>

<details>
<summary>
*More advanced (as reference, or if someone asks): perform previous 3 lines of codes as a one-liner:*
</summary>

The **tidyverse** provides a function **%>%** which **pipes** the output of one function as the input of the next function. This way, different functions from the tidyverse can be linked together into a clean piece of code.

If we want to produce a one-liner out of the 3 lines of code of the exercise, we can do it the following way:

```{r}
gtf_final <- rename(gtf, chromosome=chr) %>%
  filter(strand=="+" & chromosome=="chr4") %>%
  select(gene_symbol, gene_type)
```

If you want to learn more about the **pipe**, you can for example refer to [that page](https://style.tidyverse.org/pipes.html).

</details>
