# Data filtering

We will now see a couple of basic functions to filter and select a data object.

## filter()

**filter()** will filter **rows**.

If we take the **geneexp** objet as an example, we can filter only down-regulated genes using the **filter()** function of **{dplyr}**.

In case you need to import the data again:

```{r, eval=FALSE}
geneexp <- read_csv("DataViz_source_files-main/files/expression_20genes.csv")
```

```{r}
filter(geneexp, DE=="Down")
```

**==** is a **logical operator** that represens equality. It means that filter will return **rows in geneexp that are exactly equal to "Down"**.

Logical operators:

| Operator  | Description   |
|:--------: |:-------------:|
|     <     | less than    |
|     <=    | less than or equal to   |
|     >     | greater than  |
|     >=    | greater than or equal to|
|     ==    | exactly equal to   |
|     !=    | not equal to |
|    !x     | not x |
|    x \| y  | x OR y |
|    x & y  | x AND y |

We can have several conditions.

For example, we may want to extract only **Up** or **Down** rows from geneexp column **DE**:

```{r}
filter(geneexp, DE=="Down" | DE=="Up")
```

Here, we introduce another operator, **|**, which mean **OR**, so rows will be kept if there is either **Down** or **Up** in **DE** column.

A good practice is to assign the filtered output to a new object, for example:

```{r}
geneexp_filt <- filter(geneexp, DE=="Down" | DE=="Up")
```

## select()

**select()** will select **columns**.

Select only columns **Gene** and **DE** from **geneexp**:

```{r}
select(geneexp, Gene, DE)
```

Again, a good practice is to save that subset into a new object:

```{r}
geneexp_selected <- select(geneexp, Gene, DE)
```

## rename()

You can easily rename a column using the **rename()** function.

For example, we can rename column **Gene** to **Symbol**, the following way:

```{r}
rename(geneexp, Symbol=Gene)
```

The first argument is the data, then comes the column(s) to rename in the format: *new_name=old_name*.


## Exercise 3

We will start from **gtf** object.

Read it as follows, if it is not in your environment:

```{r, eval=F}
gtf <- read_csv("DataViz_source_files-main/files/gencode.v44.annotation.csv")
```

1. Rename column **chr** to **Chromosome**. Assign to a new object.

<details>
<summary>
correction
</summary>

```{r}
gtf2 <- rename(gtf, chromosome=chr)
```

</details>

<br>
2. Keep only rows corresponding to the **+** strand and to chromosome **chr4**. assign to a new object.

<details>
<summary>
correction
</summary>

```{r}
gtf3 <- filter(gtf2, strand=="+" & chromosome=="chr4")
```

</details>

<br>

3. Select only columns **gene_symbol** and **gene_type**.

<details>
<summary>
correction
</summary>

```{r}
gtf4 <- select(gtf3, gene_symbol, gene_type)
```

</details>

<br>

## The pipe operator

The **tidyverse** provides a function **%>%** which **pipes** the output of one function as the input of the next function. 

This way, different functions from the tidyverse can be linked together into a clean piece of code.

If we want to produce a one-liner out of the 3 lines of code of the previous exercise, we can do it the following way:

```{r}
gtf_final <- rename(gtf, chromosome=chr) %>%
  filter(strand=="+" & chromosome=="chr4") %>%
  select(gene_symbol, gene_type)
```

*Note that, in filter and select, you do not need to specify the first parameter (the data), as it is automatically looking for the output of the previous one!*

If you want to learn more about the **pipe**, you can for example refer to [that page](https://style.tidyverse.org/pipes.html).

Note that we can also link the data manipulation output to ggplot: the **output of the last command will then be used as an input to ggplot**:

```{r}
rename(gtf, chromosome=chr) %>%
  filter(strand=="+" & chromosome=="chr4") %>%
  select(gene_symbol, gene_type) %>% 
  ggplot(aes(x=gene_type)) + geom_bar()
```


## From wide to long format

The wide format is what you would typically have in a table with measurements, such as genes in rows and samples in columns.

However, we have seen that ggplot2 requires data to be **converted to a long format**.

In a long format, **one row corresponds to one observation/measurement, with all information associated to it**.

**{tidyr}** provides **pivot_longer()** to convert wide to long format, and **pivot_wider()** to convert a long to a wide format.

Our object geneexp is in a wide format. Two columns contain values of expression:

```{r, echo=F}
head(geneexp, 5)
```

In order to convert to a long format we will create:

* One column that contains the sample names
* One column that contains the expression values

```{r}
pivot_longer(geneexp, cols=c("sample1", "sample2"))
```

We can specify the names of the new columns as we create them:

```{r}
pivot_longer(geneexp, cols=c("sample1", "sample2"), names_to = "samples", values_to = "expression")
```

## Exercise 4

Let's do a small exercise to practice this conversion.

1. Read in file **DataViz_source_files-main/files/stats_countries_barcelona_2013-2023.csv**

<details>
<summary>
correction
</summary>

```{r}
stats_countries <- read_csv("DataViz_source_files-main/files/stats_countries_barcelona_2013-2023.csv")
```

</details>

<br>
2. Columns 2013 to 2022 contain **observations**. Convert this wide format into a long format. Save into a new object.

<details>
<summary>
correction
</summary>

```{r, eval=TRUE}
# Option 1: list all columns
stats_countries_long <- pivot_longer(stats_countries, 
                                     cols=c("2013", "2014", "2015", "2016", "2017", "2018", "2019", "2020", "2021", "2022", "2023"),
                                     values_to = "Population",
                                     names_to = "Years")

# Option 2: select columns that start with "20"
stats_countries_long <- pivot_longer(stats_countries, 
                                     cols=starts_with("20"),
                                     values_to = "Population",
                                     names_to = "Years")

# Option 3: select all numeric columns
stats_countries_long <- pivot_longer(stats_countries, 
                                     cols=where(is.numeric),
                                     values_to = "Population",
                                     names_to = "Years")
```

</details>

<br>
3. Do the 3 following tasks linked by a **%>%** (pipe):

* Keep only rows that correspond to the American continent and that correspond to a population of at least 50 people.
* Remove column **Continent**.
* Produce a barplot with "Years" in the x-axis, Population in the y-axis, and split by Country.

<details>
<summary>
correction
</summary>

```{r, fig.width=15}
filter(stats_countries_long, Continent=="America" & Population >= 50) %>%
  select(-Continent) %>%
  ggplot(mapping=aes(x=Years, y=Population, fill=Country)) + geom_bar(stat = "identity", position = "dodge")
```

</details>
