# ggplot2

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=5, fig.height=4,
                      echo=TRUE, warning=FALSE, message=FALSE)
```

* Graphing package inspired by the **G**rammar of **G**raphics work of Leland Wilkinson.

* Allows to compose graphs by combining independent **components / blocks / layers**.

* Flexible, highly customizable.

* Well documented.

<img src="images/ggplot_plots_cscherer.png" alt="import zip" width="700"/>

*image from https://www.cedricscherer.com/img/ggplot-tutorial/overview.png*


## Getting started

* All ggplots start with a **base layer** created with the **ggplot()** function. This is the base of what we will see during the course.

```{r, eval=F, echo=TRUE}
ggplot(data=dataframe, mapping=aes(x=column1, y=column2))
```

The function is divided into 2 main parts:

* **data** : object that contains the data that we want to graph.
* **mapping=aes()** : the "aesthetics" function tells ggplot what we want to represent in the graph. These are **data-dependent**:
  * here, we want to show *column1* (a column in **data**) on the x-axis, and *column2* (another column in **data**) on the y-axis.

*The base layer is setting the grounds but NOT producing any plot yet*.

* You need to **add a layer** (with the **+** sign) to tells R what kind of plot you want to produce out of the data.

We will now see some examples.

## Scatter plot

### Base plot

We can start from the **geneexp** object, that holds the content of file *expression_20genes.csv*: we want to plot **sample1** on the x axis and **sample2** on the y axis.

The base layer will be the following:

```{r, eval=F}
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2))
```

Adding to this the layer / block called **geom_point** that will **tell ggplot to produce a scatter/point plot**:

```{r, eval=T}
# This line is a comment: a comment is not interpreted by R.
# Example of a scatter plot: add the geom_point() layer
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2)) + 
  geom_point()

# Note that the new line is not necessary here
```

Your plot should appear in the "Plots" tab in the bottom-right panel.

### Customize the points

**geom_point()** also takes some parameters, including the color of the point, and their size:

Color all points in red:

```{r}
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2)) + 
  geom_point(color="red")
```

Make points larger (size default is 1.5):

```{r}
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2)) + 
  geom_point(color="red", size=2.5)
```

You can also have different shapes for the points!

This is done by setting the **shape** parameter in **geom_point()**.

For example, triangles:

```{r}
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2)) + 
  geom_point(color="red", size=2.5, shape="triangle")
```

<img src="images/ggplot2_shape.png" alt="import zip" width="700"/>

*Image from ggplot2 documentation*

### Add more layers

We can add more layers to the plot, using the same structure!

#### ggtitle()

For example, let's add a title using the **ggtitle()** layer:

```{r}
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2)) + 
  geom_point(color="red", size=2.5, shape="diamond") +
  ggtitle(label="my first ggplot")

```

**label** is a parameter of **ggtitle()** function.

#### Background

Not a big fan of the grey background?
This is the default "theme", but there are [more options](https://ggplot2.tidyverse.org/reference/ggtheme.html).

For example:

```{r}
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2)) + 
  geom_point(color="red", size=2.5, shape="diamond") +
  ggtitle(label="my first ggplot") +
  theme_linedraw()
```

## Save your plot

Before we dive into more graph types, let's save the current plot.

In the "Plots" tab, click on "Export" and "Save as image":

<img src="images/ggplot_save_image.png" alt="import zip" width="700"/>

From that windows, you can:

* Pick an image format between: PNG, JPEG, TIFF, BMP, SVG, EPS.
* Choose where you want to **save the output file**
* Choose the **file name**.
* Set the dimensions, by either:
  * Setting the Width and Height of the figure (in Pixels)
  * Moving the graph manually (bottom-right corner of the plot) until you obtain the size and proportions that you want.

## Exercise 1

Time for our first exercise! 

Starting from the same object **genexp**:


1. Create a scatter plot that shows sample2 on the x-axis and sample1 on the y-axis.

<details>
<summary>
correction
</summary>

```{r}
ggplot(data=geneexp, mapping=aes(x=sample2, y=sample1)) + 
  geom_point()
```

</details>

<br>
2. Change the point color to blue, and the point size to 2.

<details>
<summary>
correction
</summary>

```{r}
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2)) + 
  geom_point(color="blue", size=2)
```

</details>

<br>
3. Change the point shape to "square cross"

<details>
<summary>
correction
</summary>

```{r}
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2)) + 
  geom_point(color="blue", size=2, shape="square cross")
```

</details>

<br>
4. Add the title of your choice.

<details>
<summary>
correction
</summary>

```{r}
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2)) + 
  geom_point(color="blue", size=2, shape="square cross") +
  ggtitle(label="my second ggplot")
```

</details>

<br>
5. Add a subtitle (wait: that's new! Google "ggtitle subtitle" and see if you can find out!)

<details>
<summary>
correction
</summary>

```{r}
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2)) + 
  geom_point(color="blue", size=2, shape="square cross") +
  ggtitle(label="my second ggplot", subtitle="nice blue squares")
```

</details>

<br>
6. Save your plot as a JPEG file, in the course folder, with dimensions 600X600.

<details>
<summary>
correction
</summary>

Bottom-right pannel -> Plots tab -> Export -> ...

</details>

## Scatter plots: more features

Let's customize our scatter plot a bit more.

We may want to show the gene names that correspond to those points.

This is done by setting the "label" parameter, in the ggplot **aes()** function, and by adding the **geom_text()** layer:

```{r}
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2, label=genes)) + 
  geom_point() +
  geom_text()
```

Not bad, but let's adjust the location of the labels, so they do not overlap with the points.

We can set the **nudge_x** parameter, to move the labels horizontally (on the **x** axis).

```{r}
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2, label=genes)) + 
  geom_point() +
  geom_text(nudge_x=1.5)
```

We can also decrease or increase the label size:

```{r}
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2, label=genes)) + 
  geom_point() +
  geom_text(nudge_x=1.5, size=3)
```

Last but not least: point color can be **dependent on another column / variable of the data**.

The column we want to use to color the points should be specified inside the **aes()** function:

```{r, fig.width=7}
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2, label=genes, color=DE)) + 
  geom_point() +
  geom_text(nudge_x=1.2, size=3)
```

Colors can be set manually using (yet another) layer: **scale_color_manual()**.

This is becoming more complex in terms of code structure:

```{r, fig.width=7}
ggplot(data=geneexp, mapping=aes(x=sample1, y=sample2, label=genes, color=DE)) + 
  geom_point() +
  geom_text(nudge_x=1.2, size=3) +
  scale_color_manual(values=c("blue", "black", "red"))
```

## Bar plots

A bar plot (or bar chart) is a graph that represents categorical data with rectangular bars, which heights are proportional to the values they represent.

The first layer of the **ggplot()** function will be similar. However, note that we start with only the **x=** parameter set in **aes()** function (this is the basic way to plot a barplot):

```{r, eval=F}
ggplot(data=dataframe, mapping=aes(x=column1)) 
```

```{r, eval=F}
ggplot(data=dataframe, mapping=aes(x=column1)) +
  geom_bar()
```

Using our previous **geneexp** data, we can produce a bar plot out of the **DE** column, such as:

```{r}
ggplot(geneexp, aes(x=DE)) + 
  geom_bar()
```

This produces a barplots containing 3 bars: **Down**, **No** and **Up**: their height represents the number of genes found in each category.


### Exercise 2

1. Import file **DataViz_source_files-main/files/gencode.v44.annotation.csv** in R, into an object called **gtf**.

<details>
<summary>
correction
</summary>

```{r}
gtf <- read_csv("DataViz_source_files-main/files/gencode.v44.annotation.csv")
```

This is a subset of the gencode v44 human gene annotation:

* Only protein coding, long non-coding, miRNAs, snRNAs and snoRNAs
* Limited to chromosomes 1 to 10
* Random subset of 1000 genes
* Converted to a friendly csv format.

</details>


2. Create a simple barplot representing the count of genes per chromosome:

<details>
<summary>
correction
</summary>

```{r}
ggplot(data=gtf, mapping=aes(x=chr)) + 
  geom_bar()
```

</details>


3. Keeping the chromosome on the x axis, split the barplot **per gene type**.

TIP: remember how we set **color=** in **mapping=aes()** function in the scatter plot section?

<details>
<summary>
correction
</summary>

```{r}
ggplot(data=gtf, mapping=aes(x=chr, color=gene_type)) + 
  geom_bar()
```

</details>

4. Now try with **fill** instead of **color**:

<details>
<summary>
correction
</summary>

```{r}
ggplot(data=gtf, mapping=aes(x=chr, fill=gene_type)) + 
  geom_bar()
```

</details>

5. Add a title to the graph:

<details>
<summary>
correction
</summary>

```{r}
ggplot(data=gtf, mapping=aes(x=chr, fill=gene_type)) + 
  geom_bar(position="dodge") +
  ggtitle(label = "Number of genes per chromosome, split by gene type")
```

</details>

6. Change the **theme**:

<details>
<summary>
correction
</summary>

```{r}
ggplot(data=gtf, mapping=aes(x=chr, fill=gene_type)) + 
  geom_bar(position="dodge") +
  ggtitle(label = "Number of genes per chromosome, split by gene type") +
  theme_bw()
```

</details>

7. Save graph in PNG format in the course's directory.

### Bars positions

We can also play with the **position** of the bars. By default, position is **stack**, i.e. categories are stacked on top of each other along the bar.

Position **fill** will show the proportions, instead of the absolute values, of each category:

```{r}
ggplot(data=gtf, mapping=aes(x=chr, fill=gene_type)) + 
  geom_bar(position="fill")
```

Position **dodge** is representing each category (here, continents), next to each other:

```{r}
ggplot(data=gtf, mapping=aes(x=chr, fill=gene_type)) + 
  geom_bar(position="dodge")
```




<details>
<summary>
*More advanced (as reference, or if someone asks): how to reorder x-axis labels:*
</summary>

```{r}
ggplot(data=gtf, mapping=aes(x=factor(chr, levels=c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9", "chr10"), ordered=T), fill=gene_type)) + 
  geom_bar(position="dodge") +
  xlab("chromosome")
```

</details>



### stats="identity" parameter

We will now see a different type of data R can produce a barplot from.

Let's import data from file: *DataViz_source_files-main/files/stats_continents_barcelona_2013-2023_long.csv* 
into an object called **statsbcn**.

The data contains the number of foreign residents in Barcelona from 2013 to 2023.

```{r}
statsbcn <- read_csv("DataViz_source_files-main/files/stats_continents_barcelona_2013-2023_long.csv")
```

How many rows and how many columns does the data contain?


In the barplots we created so far, R takes categories in the columns specified **x=** and counts the number of occurrences.

The argument **stat="identity"** in geom_bar() is telling R to calculate the sum of the **y** variable, **grouped by the x variable**: bars of the barplot will display the sums.

In the following example, we are plotting the sum of foreign residents in Barcelona (Population provided in **y**) per year (Year provided in **x**):

```{r}
ggplot(statsbcn, aes(x=Year, y=Population)) + 
  geom_bar(stat="identity")
```


Here, we would provide **Continent** to **fill**:

```{r}
ggplot(statsbcn, aes(x=Year, y=Population, fill=Continent)) + 
  geom_bar(stat="identity")
```

We can here again play with the **position**. 

Position **fill** :

```{r}
ggplot(statsbcn, aes(x=Year, y=Population, fill=Continent)) + 
  geom_bar(stat="identity", position="fill")
```

Position **dodge** :

```{r}
ggplot(statsbcn, aes(x=Year, y=Population, fill=Continent)) + 
  geom_bar(stat="identity", position="dodge")
```



<details>
<summary>
*More advanced (as reference, or if someone asks): display all labels:*
</summary>

Convert "Year" column as character, instead of numbers:

```{r}
ggplot(statsbcn, aes(x=as.character(Year), y=Population, fill=Continent)) + 
  geom_bar(stat="identity", position="dodge")
```

</details>






