# More ggplots

## Histograms and density plots

### Histogram

We can use one of our previous objects to create a histogram. You can get it back with:

```{r}
rnaseq2 <- read_csv("DataViz_source_files-main/files/GSE150029_rnaseq_log2_long.csv")
```

```{r}
ggplot(rnaseq2, aes(x=log2_counts)) + 
  geom_histogram()
```

We can split the histogram per sample:

```{r}
ggplot(rnaseq2, aes(x=log2_counts, color=sample)) + 
  geom_histogram() 
```

Set position to "identity", so histograms will not be on top of each other:

```{r}
ggplot(rnaseq2, aes(x=log2_counts, color=sample)) + 
  geom_histogram(position="identity")
```

Set alpha (transparency to 0.2):

```{r}
ggplot(rnaseq2, aes(x=log2_counts, color=sample)) + 
  geom_histogram(position="identity", alpha=0.2)
```

Express as density instead of counts:

```{r}
ggplot(rnaseq2, aes(x=log2_counts, color=sample)) + 
  geom_histogram(aes(y=after_stat(density)), position="identity", alpha=0.2) +
  theme_classic()
```


### Density plot

The structure is pretty much the same:

```{r}
ggplot(rnaseq2, aes(x=log2_counts, color=sample)) + 
  geom_density(alpha=0.2) +
  theme_classic()
```

You can use fill instead:

```{r}
ggplot(rnaseq2, aes(x=log2_counts, fill=sample)) + 
  geom_density(alpha=0.2) +
  theme_classic()
```

### Histogram + density

As we combined geom_boxplot and geom_violin in a previous exercise, we can also combine geom_histogram and geom_density:

```{r}
ggplot(rnaseq2, aes(x=log2_counts, color=sample)) + 
  geom_histogram(aes(y=after_stat(density)), position="identity", alpha=0.2) +
  geom_density(alpha=0.2) +
  theme_classic()
```


## Pie chart

We will plot a pie chart representing the number of genes from **gtf** found in each chromosome.

Prepare data so as to obtain counts (that is more advanced dplyr manipulation):

```{r}
gtf_count_chr <- gtf %>%
  group_by(chr) %>%
  summarise(counts=n())
```

The **coord_polar()** is needed here: pie plots are stacked barplots in polar coordinates.

```{r}
ggplot(data=gtf_count_chr, mapping=aes(x="", y=counts, fill=chr)) + 
  geom_bar(stat="identity") +
  coord_polar("y", start=0)
```

```{r}
ggplot(data=gtf_count_chr, mapping=aes(x="", y=counts, fill=chr)) + 
  geom_bar(stat="identity") +
  coord_polar("y", start=0) +
  theme_void()
```

Change the color scale:

```{r}
ggplot(data=gtf_count_chr, mapping=aes(x="", y=counts, fill=chr)) + 
  geom_bar(stat="identity") +
  coord_polar("y", start=0) +
  scale_fill_brewer(palette="Paired") +
  theme_void()
```

Here is a [nice page](https://genchanghsu.github.io/ggGallery/posts/2021-06-18-post-3-pie-charts-with-ggplot/) about more pie plot customization.


## Custom scatter plot

Read in data sent by Laura:

```{r}
orgs <- read_csv("DataViz_source_files-main/files/Dados_organoids.csv")
```

The idea is to visualize the "Pas" on the x-axis and the samples (column "name") on the y-axis, while controling size and color from another column.

We need to convert this to a "long" format so as to move all "Pas" columns into one:

```{r}
orgs_long=pivot_longer(orgs, cols = starts_with("Pas"), names_to="Pas", values_to = "Expr")
```

Then we build a scatter plot, based on this new long object:

```{r, fig.width=6, fig.height=5}
ggplot(orgs_long, aes(x=Pas, y=name, size=as.character(Size), color=Color)) + 
  geom_point() + 
  scale_color_manual(values=c(Blue="lightblue", Red="red")) + 
  theme_minimal() + 
  theme(panel.grid.major.y = element_line(color="grey82", arrow = arrow(angle = 25, length = unit(.15, "inches"),type = "open")))
```

Remove the legend and the axis names:

```{r, fig.width=5, fig.height=5}
ggplot(orgs_long, aes(x=Pas, y=name, size=as.character(Size), color=Color)) + 
  geom_point() + 
  scale_color_manual(values=c(Blue="lightblue", Red="red")) + 
  theme_minimal() + 
  theme(panel.grid.major.y = element_line(color="grey82", arrow = arrow(angle = 25, length = unit(.15, "inches"),type = "open")), legend.position="none", axis.title = element_blank())
```
